open-apparel-registry/open-apparel-registry
# Redesign facility grid layer

## Overview

Incorporate design feedback into a revised facility grid layer that uses a hexagonal grid and circular symbology rather than a square grid.

Connects #768

## Demo

![2019-09-10 10 04 30](https://user-images.githubusercontent.com/17363/64635095-b6251100-d3b3-11e9-9347-15585a4dd3e4.gif)


## Notes

We implemented the first grid layer to aggregate facilities at low zoom levels in #755. The general consensus was that the grid was revealing interesting patters in the data, but had some problems.

- The sparseness of the grid made the layer look broken, as if it was not finished loading.
- It was not obvious that the grid squares were interactive.
- Distortion of the squares at very low zoom levels.

Some potential modifications were discussed.

  - Show empty grid cells with a faint border. Clip them to land masses.
  - Use circles instead of squares.
  - Use different color ramps.
  - Use a hex grid instead of a square grid.

I added the empty grid cells with a faint border since it was quick to do, but did not go to the trouble of loading land mass boundaries and modifying the query to clip the data. It looked very busy just over land and did not address the other shortcomings of the grid approach.

Implementing circular symbology was also relatively straightforward and produced much more pleasing results.

<img width="1121" alt="Screen Shot 2019-09-06 at 1 48 27 PM" src="https://user-images.githubusercontent.com/17363/64637437-a6f49200-d3b8-11e9-8404-cf3ed2dea031.png">
 
In order to remove as much of the gap between the circles as possible I researched generating a hex grid rather than a square grid and found [an easy-to-use postgres function](https://gist.github.com/mjumbewu/1761802ea06fb78c596f9cf8c9b2e769) to build it. Using a function also eliminates an inefficiency in our grid approach, where we generated and sent a large SQL string to construct the grid for each tile. The tradeoff for using the hex grid was the need to calculate the bounds of each feature rather than use a ZXY triple, but the benefits far outweighed the cost.

## Testing Instructions

* Run `./scripts/update` to ensure the required migration is run
* Run `./scripts/manage dbshell` and verify the current version number. 
```
openapparelregistry=# SELECT version FROM api_version;
 version
---------
       8
```
* Run `./scripts/manage incrementtileversion` to ensure the cache is invalidate.
* Run `./scripts/manage dbshell` and verify the current version number has been incremented. 
```
openapparelregistry=# SELECT version FROM api_version;
 version
---------
       9
```
* Browse the and verify that grid circles display a color and size relative to the number of points they represent.
* Verify that clicking a circle zooms to the extent.
* Verify that the grid is filtered properly when searching. 

## Checklist
